tests.md

```diff
@@ -0,0 +1,174 @@
+# Testing of Terraform Modules Explained
+
+## Are Unit Tests Viable?
+
+A useful definition of a unit test is one which mainly spends its run time in the code that we own, and very little
+time in the code owned by other organizations or teams. There are other definitions, but the usefulness of this
+is dictated by tightening the loop of developer's work. The progress is faster with a five-seconds test run than with
+a five-hours test run as the developer waits for the result to decide whether to go further with the work or to go
+back and correct the last portion of their work. The more time the test run uses for the others' code, the less time
+it spends inside our code, to catch the kind of errors *this team* is able to fix by ourselves.
+
+How does that translate to the world of Terraform reusable modules? The full Terraform run which would use the code
+of a reusable Terraform module always uses some provider, and the provider always speaks to some cloud's REST API.
+Typically the provider+cloud run time is much longer, orders of magnitude longer, than the module run time. Realistic
+example may be 100 seconds versus less than 1 second.
+
+To enable running multiple test cases in much shorter time than that, one can either mock the provider
+or mock the cloud's REST API. Such a "mocked" component basically either returns hardcoded results or an error
+when anything unexpected happens, so it only takes microseconds or so. However such method requires either:
+
+- the Terraform team to be versed in Go language, to be able to implement their own entire mock provider,
+- or, alternatively, the Terraform team to be versed enough in some language to implement their own cloud REST API,
+to have intimate knowledge of the original REST API of that cloud, and to be able to configure the provider to not to
+speak to the real cloud anymore.
+
+Such large investment appears for now to be only possible through an organization-level decision, not a team-level one.
+So the default is not to do the testing this way. Instead we resort to mainly `terraform apply` to the real cloud,
+which one would normally call integration testing. Just as for example Go/Python projects' integration tests which could
+speak to the third-party API or spin up the database instance.
+
+As a corollary to this, the terms "Unit Test" and "Integration Test" should be used with care with Terraform. If you
```

Comment:
https://www.hashicorp.com/blog/testing-hashicorp-terraform

Reply:
Please clarify? Do you mean to use that blog entry as a definite reference and modify this section to be less speculative in tone and a more firm? ("howto" mode)

Reply:
I've removed the speculative bits.